hbase-client/src/main/java/org/apache/hadoop/hbase/client/RegionInfo.java

```diff
@@ -69,7 +65,14 @@
  */
 @InterfaceAudience.Public
 public interface RegionInfo extends Comparable<RegionInfo> {
-  RegionInfo UNDEFINED = RegionInfoBuilder.newBuilder(TableName.valueOf("__UNDEFINED__")).build();
+  /**
+   * Do not use.
+   * @deprecated Since 2.3.1; to be removed in 4.0.0 with no replacement.
+   */
+  // Removed because creation was creating a static deadlock, HBASE-24896
+  @Deprecated
+  RegionInfo UNDEFINED = null;
```

Comment:
Was wondering if we can move the following into RegionInfo to avoid this deprecation stuff. For that we will have to switch to RegionInfoBuilder c'tor because the following usage is a private c'tor. Curious if you considered that and didn't want to do it for some reason. The c'tor was specifically designed for this usecase, so I may be missing some context here.

public static final RegionInfo FIRST_META_REGIONINFO =
    new MutableRegionInfo(1L, TableName.META_TABLE_NAME, RegionInfo.DEFAULT_REPLICA_ID);

Reply:
Thanks for the review @bharathv 

Say more. You suggesting move FIRST_META_REGIONINFO define into RegionInfo?

Reply:
> You suggesting move FIRST_META_REGIONINFO define into RegionInfo?

Exactly. I think thats one way to break the loop without this deprecation schedule. I don't know if it has other implications (especially if it breaks the semantics described in HBASE-17980).

Reply:
I'd glanced in passing but there are ~100 refs to old location. Let me try it... make sure we are not just moving this problem. Would be better I agree if it worked and we could avoid this deprecation. @bharathv